### ISSUE TITLE
Why is ODataUrlKeyDelimiter.Slash the default instead of ODataUrlKeyDelimiter.Parenthesis?

### ISSUE THREAD

**Issue by bdebaere** [OData/AspNetCoreOData#813]
According to the OData standard [Addressing Entities](https://docs.oasis-open.org/odata/odata/v4.01/odata-v4.01-part2-url-conventions.html#sec_AddressingEntities), entities should be addressed with parenthesis.
Does the OData standard dictate whether I should support parenthesis, slashes or both, i.e. should I switch my routes to `[HttpGet("odata/Products/{key}")]` instead of `[HttpGet("odata/Products({key})")]` or allow both?
Since alternate keys, and composite keys, are only really possible with parenthesis (correct me if I'm wrong), and we are using these e.g. `[HttpGet("odata/Products(number={number})")]`, it seems to me we should be consistent and support parenthesis in our other methods including `odata/Products(number='123')/Attributes`. So at the very least I would add both routes to every method except the alternate keys where (correct me if I'm wrong) there's no choice.

**Comment by julealgon** [comment]
There are a few settings related to parenthesis ID routes in the OData configuration in startup. Did you try playing around with them?

**Comment by bdebaere** [comment]
@julealgon Thanks a lot! Your post reminded me of `options.UrlKeyDelimiter = ODataUrlKeyDelimiter.Parentheses;`. I'm still curious why this is not the default though and it is set to `public ODataUrlKeyDelimiter UrlKeyDelimiter { get; set; } = ODataUrlKeyDelimiter.Slash;` in `ODataOptions.cs`.

**Comment by julealgon** [comment]
> I'm still curious why this is not the default though and it is set to `public ODataUrlKeyDelimiter UrlKeyDelimiter { get; set; } = ODataUrlKeyDelimiter.Slash;` in `ODataOptions.cs`.

I'd assume an attempt to standardize across normal MVC routes? I've always found weird the way OData managed the ID segment without a segment but with parenthesis. I think this was then later relaxed in the spec to allow both.

> My question isn't what's possible, my question is: what is advised based on the OData spec/standard?

This however is still a good question that I'm also interest to hear the response for. I do agree with you that using slash kind of introduces a potential nasty inconsistency with composite key routes, which would require you to specify IDs in different ways in the same API.

**Comment by bdebaere** [comment]
Maybe @mikepizzo has an answer on how one can make their OData API conform to the standards while keeping in mind alternate keys and @odata.nextlink? Why isn't `public ODataUrlKeyDelimiter UrlKeyDelimiter { get; set; } = ODataUrlKeyDelimiter.Parenthesis;` the standard in `ODataOptions.cs`?

**Comment by xuzhg** [comment]
> s however is still a good question that I'm also interest t

I remembered in the early 8 design and implementation, Mike and I thought "key as segment (aka, Slash)" is more popular in building REST APIs. It could be the reason.

In 8, both `[HttpGet("odata/Products/{key}")]` and `[HttpGet("odata/Products({key})")]` should work. Because you are using attribute routing. You can choose to decorate one, the other, or both. Either way should work.

In 8 conventional routing, if the action/controller meets the convention rule, i will try to build routes as many as possible. 
So,`odata/Products/{key}` and `odata/Products({key})` are for a single key. 

but, one `odata/Products({key})` for composite key and alternate key. If my memory is correct. :)

I'd like hear @mikepizzo 's input.

**Comment by mikepizzo** [comment]
The key-as-segment/parens-key argument has been around as long as OData has existed. In OData v1 we chose to go with the parens-syntax since it was unambiguous and extensible to composite and alternate keys. From the time we published OData v1 we got push back that API designers didn't like to put parens in their URLs, and most REST APIs designed without being well structured around a model with composable semantics used a slash before a key value within a collection.  So, we decided to support key-as-segment as an alternate syntax, but continued to require parens-style syntax for interoperability.

Supporting key-as-segment required that we define precedence rules for resolving potential ambiguities. The 4.01 specification does suggest that you could support key-as-segment for multi-part keys by ordering keys in the url in the same order as they appear in metadata, but it doesn't answer questions like what happens if you only specify one of the multi-part segments, and it does not (nor are there plans to) define semantics for alternate keys using key-as-segment.

The parens-key syntax remains the canonical snytax for an entity, and parens are used within the context url.

The contents of nextLink and deltaLink, as well as things like navigation links where included in the payload, are defined to be opaque to the client. That is, the service can choose to use key-as-segment, parens segment, or something completely random, and it shouldn't matter to the client.  However, it is convenient to model these after equivalent OData conventions, so they typically follow either key-as-segment or parens-key syntax.

Bottom line; when implementing an API I would support both parens-key syntax and, for single part primary keys (which are the common case), key as segment. When consuming an API, I would use the parens-key syntax as it is unambiguous and supports multi-part and alternate keys.
